src/dbnode/storage/bootstrap/bootstrapper/peers/source.go

```diff
@@ -617,98 +626,141 @@ func (s *peersSource) readIndex(
 	// FOLLOWUP(r): Try to reuse any metadata fetched during the ReadData(...)
 	// call rather than going to the network again
 	r := result.NewIndexBootstrapResult()
+	builders := make(result.IndexBuilders)
 	if shardsTimeRanges.IsEmpty() {
 		return r, nil
 	}
 
-	session, err := s.opts.AdminClient().DefaultAdminSession()
-	if err != nil {
-		s.log.Error("peers bootstrapper cannot get default admin session", zap.Error(err))
-		r.SetUnfulfilled(shardsTimeRanges)
-		return nil, err
-	}
-
 	var (
-		count         = len(shardsTimeRanges)
-		concurrency   = s.opts.DefaultShardConcurrency()
-		dataBlockSize = ns.Options().RetentionOptions().BlockSize()
-		resultOpts    = s.opts.ResultOptions()
-		idxOpts       = ns.Options().IndexOptions()
-		resultLock    = &sync.Mutex{}
-		wg            sync.WaitGroup
+		count          = len(shardsTimeRanges)
+		concurrency    = s.opts.DefaultShardConcurrency()
+		indexBlockSize = ns.Options().RetentionOptions().BlockSize()
+		runtimeOpts    = s.opts.RuntimeOptionsManager().Get()
+		resultOpts     = s.opts.ResultOptions()
+		fsOpts         = s.opts.FilesystemOptions()
+		idxOpts        = ns.Options().IndexOptions()
+		readerPool     = bootstrapper.NewReaderPool(bootstrapper.NewReaderPoolOptions{
+			Alloc: func() (fs.DataFileSetReader, error) {
+				bytesPool := s.opts.ResultOptions().DatabaseBlockOptions().BytesPool()
+				return fs.NewReader(bytesPool, fsOpts)
+			},
+		})
+		resultLock                  = &sync.Mutex{}
+		readersCh                   = make(chan bootstrapper.TimeWindowReaders)
+		shouldPersistIndexBootstrap = true
+		wg                          sync.WaitGroup
 	)
 	s.log.Info("peers bootstrapper bootstrapping index for ranges",
 		zap.Int("shards", count),
 		zap.Int("concurrency", concurrency),
 	)
 
+	go bootstrapper.EnqueueReaders(ns, opts, runtimeOpts, fsOpts, shardsTimeRanges, readerPool,
+		readersCh, shouldPersistIndexBootstrap, indexBlockSize, s.log)
+
 	workers := xsync.NewWorkerPool(concurrency)
 	workers.Init()
-	for shard, ranges := range shardsTimeRanges {
-		shard, ranges := shard, ranges
+	for timeWindowReaders := range readersCh {
+		timeWindowReaders := timeWindowReaders
 		wg.Add(1)
+		// NB(bodu): This is fetching the data for all shards for a block of time.
 		workers.Go(func() {
-			defer wg.Done()
-
-			iter := ranges.Iter()
-			for iter.Next() {
-				target := iter.Value()
-				size := dataBlockSize
-				for blockStart := target.Start; blockStart.Before(target.End); blockStart = blockStart.Add(size) {
-					currRange := xtime.Range{
-						Start: blockStart,
-						End:   blockStart.Add(size),
+			var (
+				docsPool                   = s.opts.IndexOptions().DocumentArrayPool()
+				batch                      = docsPool.Get()
+				indexBlockDocumentsBuilder segment.DocumentsBuilder
+				timesWithErrors            []time.Time
+			)
+			defer func() {
+				docsPool.Put(batch)
```

Comment:
nit: Want to put the `wg.Done()` in here too? (just in case an early return get's added to this function body)